lib/pbench/cli/server/report.py

```diff
@@ -368,26 +369,78 @@ def columnize(
         click.echo(line)
 
 
-def summarize_dates(rows: Iterator[Row], width: int = 80):
+def summarize_dates(base_query: Query, options: dict[str, Any]):
+    """Collect and report statistics
+
+    The caller supplies a base query providing a "date" column, effectively
+    "SELECT dataset.upload AS date" so that we can filter on the date and
+    process each match.
+
+    Args:
+        base_query: a SQLAlchemy Query producing a "date" column
+        options: The Click option dictionary
+    """
+    width: int = options.get("width")
+    since = options.get("since")
+    until = options.get("until")
+
     by_year = defaultdict(int)
     by_month = defaultdict(int)
     by_day = defaultdict(int)
     by_weekday = defaultdict(int)
     by_hour = defaultdict(int)
 
-    day = datetime.datetime.now(datetime.timezone.utc).replace(
-        hour=0, minute=0, second=0, microsecond=0
+    start = (
+        since if since else datetime.datetime.fromtimestamp(0.0, datetime.timezone.utc)
     )
+    end = until if until else datetime.datetime.now(datetime.timezone.utc)
+
+    # It's convenient to use `--until YYYY-MM-01` to see a month (though
+    # technically that would include a YYYY-MM-01:00:00.00 timestamp), but
+    # bucketizing the day or week based on that anomaly isn't very useful, so
+    # back up the "day" one millisecond to move it into the last day of the
+    # previous month.
+    day = end - datetime.timedelta(milliseconds=1)
+    day = day.replace(hour=0, minute=0, second=0, microsecond=0)
     month = day.replace(day=1)
     year = month.replace(month=1)
     week = day - datetime.timedelta(days=7)
 
+    first: Optional[datetime.datetime] = None
+    last: Optional[datetime.datetime] = None
+
     this_year = 0
     this_month = 0
     this_week = 0
     this_day = 0
+    in_range = 0
+
+    filters = []
+
+    # Create a subquery from our basic select parameters so that we can use
+    # the label (SQL "AS date") in our WHERE filter clauses. (In a direct query
+    # PostgreSQL doesn't allow filtering on renamed columns.)
+    subquery = base_query.subquery()
+    query = Database.db_session.query(subquery.c.date)
+    if since and until:
+        if since > until:
+            raise Exception("The --until value must be later than the --since value")
```

Comment:
I would be inclined to combine the conditionals together:
```python
    if since and until and since > until:
```
which saves a level of indentation.

However, of more substance, shouldn't this check be done earlier (like before initializing `by_year` et al.)?

Reply:
While it'd be more "efficient" further up, (I was going to do that, and I don't remember what happened), it's not a big issue.